packages/tpetra/core/src/Tpetra_Distributor.hpp

```diff
@@ -1339,6 +1368,100 @@ namespace Tpetra {
+              Packet* rawRecvBuf = recvBuf.getRawPtr ();
+              const int count = as<int> (recvBuf.size ());
+              MPI_Request rawRequest = MPI_REQUEST_NULL;
+              MPI_Recv_init (rawRecvBuf, count, rawType, procsFrom_[i], tag, rawComm, &rawRequest);
+
+              Teuchos::RCP<Teuchos::CommRequest<int> > req = Teuchos::Details::mpiCommRequest(rawRequest, Teuchos::arcp_reinterpret_cast<const char>(recvBuf));
+              Teuchos::set_extra_data( recvBuf, "buffer", inOutArg(req) );
+              persistentRequests_.push_back(req);
+            }
+            else { // Receiving from myself
+              selfReceiveOffset = curBufOffset; // Remember the self-recv offset
+            }
+            curBufOffset += curBufLen;
+          }
+
+          // create the persistent send requests
+          if (verbose_) {
+            std::ostringstream os;
+            os << *prefix << ": Creating persistent send requests" << endl;
+            *out_ << os.str ();
+          }
+          for (size_t i = 0; i < numBlocks; ++i) {
+            size_t p = i + procIndex;
+            if (p > (numBlocks - 1)) {
+              p -= numBlocks;
+            }
+
+            if (procsTo_[p] != myRank) {
+
+              ArrayRCP<const Packet> sendBuf =
+                exports.persistingView (startsTo_[p] * numPackets,
+                                        lengthsTo_[p] * numPackets);
+
+              Packet t;
+              MPI_Datatype rawType = Teuchos::Details::MpiTypeTraits<Packet>::getType (t);
+              const Packet* rawSendBuf = sendBuf.getRawPtr ();
+              const int count = as<int> (sendBuf.size ());
+              MPI_Request rawRequest = MPI_REQUEST_NULL;
+              MPI_Send_init (rawSendBuf, count, rawType, procsTo_[p], tag, rawComm, &rawRequest);
+
+              Teuchos::RCP<Teuchos::CommRequest<int> > req = Teuchos::Details::mpiCommRequest(rawRequest, Teuchos::arcp_reinterpret_cast<const char>(sendBuf));;
+              Teuchos::set_extra_data( sendBuf, "buffer", inOutArg(req) );
+              persistentRequests_.push_back(req);
+            }
+          }
+        }
+
+        if (verbose_) {
+          std::ostringstream os;
+          os << *prefix << ": Starting persistent requests" << endl;
+          *out_ << os.str ();
+        }
+        int count = persistentRequests_.size();
+        Array<MPI_Request> rawMpiRequests (count, MPI_REQUEST_NULL);
+        int k = 0;
+        for (int i = 0; i<persistentRequests_.size(); i++) {
+          Teuchos::RCP<Teuchos::CommRequest<int> > request = persistentRequests_[i];
+          Teuchos::RCP<Teuchos::MpiCommRequestBase<int> > mpiRequest =
+            Teuchos::rcp_dynamic_cast<Teuchos::MpiCommRequestBase<int> > (request);
+          rawMpiRequests[k] = mpiRequest->getRawMpiRequest();
+          k++;
+        }
+        const int err = MPI_Startall (count, rawMpiRequests.getRawPtr());
```

Comment:
Why wouldn't we start each send as it's processed?  Why do we start them all at once?